internal/server/controller_e2e_test.go

```
@@ -0,0 +1,372 @@
+}
+
+// openSubscription opens a websocket connection to the GraphQL
+// endpoint and performs the graphql-transport-ws "connection_init" /
+// "connection_ack" handshake.
+func openSubscription(t *testing.T, ctx context.Context, baseURL string) (*websocket.Conn, []byte) {
+	t.Helper()
+	wsURL := strings.Replace(baseURL, "http://", "ws://", 1)
+	dialer := websocket.Dialer{
+		Subprotocols: []string{"graphql-transport-ws"},
+	}
+	conn, _, err := dialer.DialContext(ctx, wsURL, nil)
```

golangci-lint (2.12.1)

[error] response body must be closed

(bodyclose)